src/core/server/workspaces/workspace_permission_check.ts

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { cloneDeep } from 'lodash';
+
+export interface Permissions {
+  write?: Principals;
+  read?: Principals;
+  management?: Principals;
+  library_read?: Principals;
+  library_write?: Principals;
+}
+
+export interface Principals {
+  users?: string[];
+  groups?: string[];
+}
+
+export const PERMISSION_TYPE_READ = 'read';
+export const PERMISSION_TYPE_WRITE = 'write';
+export const PERMISSION_TYPE_MANAGEMENT = 'management';
+export const PERMISSION_TYPE_LIBRARY_READ = 'library_read';
+export const PERMISSION_TYPE_LIBRARY_WRITE = 'library_write';
```

Comment:
I am thinking why does ACL utility need to be aware of the permission types, actually it is a pure ACL function that only knows there are `permissions: string[]` need to be check, it should not be coupled with workspace or other business logic.

Reply:
In our design, permissions are not a string array, they are permission_type->principals map, so ACL need to use permission types.

Reply:
I think @SuZhou-Joe means should ACL functions maintain a list of `permission_type`? It seems ACL functions are `permission_type` agnostic, which means it can accept any `permission_type`. For ACL, it's just a plain string to query and match.

Reply:
Yeah, I've refactored the code yet.